builtin/logical/database/backend.go

```diff
@@ -286,3 +485,329 @@ cassandra, mssql, mysql, postgres
+		}
+	}
+	return nil
+}
+
+// findStaticWAL loads a WAL entry by ID. If found, only return the WAL if it
+// is of type staticWALKey, otherwise return nil
+func (b *databaseBackend) findStaticWAL(ctx context.Context, s logical.Storage, id string) *setCredentialsWAL {
+	wal, err := framework.GetWAL(ctx, s, id)
+	if err != nil {
+		b.Logger().Warn("error reading WAL", "id", id, "error", err)
+		return nil
+	}
+
+	if wal == nil || wal.Kind != staticWALKey {
+		return nil
+	}
+
+	var walEntry setCredentialsWAL
+	if err = mapstructure.Decode(wal.Data, &walEntry); err != nil {
+		b.Logger().Warn("error decoding walEntry", err.Error())
+		return nil
+	}
+
+	return &walEntry
+}
+
+type setStaticAccountInput struct {
+	RoleName   string
+	Role       *roleEntry
+	Password   string
+	CreateUser bool
+	WALID      string
+}
+
+type setStaticAccountOutput struct {
+	RotationTime time.Time
+	Password     string
+	// Optional return field, in the event WAL was created and not destroyed
+	// during the operation
+	WALID string
+}
+
+// setStaticAccount sets the password for a static account associated with a
+// Role. This method does many things:
+// - verifies role exists and is in the allowed roles list
+// - loads an existing WAL entry if WALID input is given, otherwise creates a
+// new WAL entry
+// - gets a database connection
+// - accepts an input password, otherwise generates a new one via gRPC to the
+// database plugin
+// - sets new password for the static account
+// - uses WAL for ensuring passwords are not lost if storage to Vault fails
+//
+// This method does not preform any operations on the priority queue. Those
+// tasks must be handled outside of this method.
+func (b *databaseBackend) setStaticAccount(ctx context.Context, s logical.Storage, input *setStaticAccountInput) (*setStaticAccountOutput, error) {
+	// lvr is the known LastVaultRotation
+	var lvr time.Time
+	var merr error
+	// re-use WAL ID if present, otherwise PUT a new WAL
+	output := &setStaticAccountOutput{WALID: input.WALID}
+
+	dbConfig, err := b.DatabaseConfig(ctx, s, input.Role.DBName)
+	if err != nil {
+		return output, err
+	}
+
+	// If role name isn't in the database's allowed roles, send back a
+	// permission denied.
+	if !strutil.StrListContains(dbConfig.AllowedRoles, "*") && !strutil.StrListContainsGlob(dbConfig.AllowedRoles, input.RoleName) {
+		return output, fmt.Errorf("%q is not an allowed role", input.RoleName)
+	}
+
+	// Get the Database object
+	db, err := b.GetConnection(ctx, s, input.Role.DBName)
+	if err != nil {
+		return output, err
+	}
+
+	// Use password from input if available. This happens if we're restoring from
+	// a WAL item or processing the rotation queue with an item that has a WAL
+	// associated with it
+	newPassword := input.Password
+	if newPassword == "" {
+		// Generate a new password
+		newPassword, err = db.GenerateCredentials(ctx)
+		if err != nil {
+			return output, err
+		}
+	}
+	output.Password = newPassword
+
+	db.RLock()
```

Comment:
What resource is this protecting?

Reply:
It's supposed to protect the connection returned from `b.GetConnection`, so that it's not cleared or removed out from under us. I think after some refactoring it mistakenly got shoved down, I've moved it to just below the `b.GetConnection` call in 447bd4de0, so that it protects the `db.GenerateCredentials` call too. Does that clarify things?